[CAMEL-12711] Add configuration property 'bindAddress' for SFTP

ffeisst opened the pull request:
Add configuration property 'bindAddress' for SFTP to specify the local interface to which the SFTP connection should bind.

Because the underlying SFTP library JSch does not support setting the bind address directly, I have configured the JSch Session with an own ConnectionFactory which opens a Socket by specifying the bindAddress.

oscerd commented:
Can you please raise a JIRA in https://issues.apache.org/jira/browse/CAMEL and reference the JIRA issue number in your commit message to track this change? Thanks.

onderson commented on components/camel-ftp/src/main/java/org/apache/camel/component/file/remote/SftpOperations.java:
```diff
@@ -1074,4 +1096,70 @@ public synchronized boolean sendSiteCommand(String command) throws GenericFileOp
+        final Socket[] sockp=new Socket[1];
+        final Exception[] ee=new Exception[1];
+        String message="";
+        Thread tmp=new Thread(new Runnable(){
```
why new runnable needed? can you please elaborate on it?

onderson replied:
Ohh i see, i found a similar piece of this. I think this is public static method, no copy paste needed i guess. it may be imported. i did not fully check..

ffeisst replied:
I wanted to stay as close as possible to the original socket creation implementation. Therefore I copied the mentioned Util method (which has package protected access, btw) and added the possibility to specify the bind address to it (the original method does not support specifying the bind address). I've also added a comment on the method with a link to the original method.

The runnable here is used to implement the timeout behavior.

onderson commented on components/camel-ftp/src/main/java/org/apache/camel/component/file/remote/SftpOperations.java:
```diff
@@ -1074,4 +1096,70 @@ public synchronized boolean sendSiteCommand(String command) throws GenericFileOp
+          }
+          tmp.interrupt();
+          tmp=null;
+          throw new RuntimeException(message, ee[0]);
```
RuntimeCamelException would be bettter..

ffeisst replied:
Thanks for the review, I've changed this accordingly.

dmvolod commented:
@ffeisst , please run maven with -Psourcecheck to validate code style

ffeisst commented:
I have raised a JIRA issue (https://issues.apache.org/jira/browse/CAMEL-12711) and updated the commit message. I've also fixed all the checkstyle warnings.

onderson commented on components/camel-ftp/src/main/java/org/apache/camel/component/file/remote/SftpOperations.java:
```diff
@@ -1074,4 +1097,64 @@ public synchronized boolean sendSiteCommand(String command) throws GenericFileOp
+     * added possibility to specify the address of the local network interface, against the
+     * connection should bind
+     */
+    static Socket createSocketUtil(final String host, final int port, final String bindAddress, final int timeout) {
```
what you do is reasonably good as the class and method has default access modifers.

onderson commented on components/camel-ftp/src/main/java/org/apache/camel/component/file/remote/SftpOperations.java:
```diff
@@ -1074,4 +1097,64 @@ public synchronized boolean sendSiteCommand(String command) throws GenericFileOp
+            public void run() {
+                sockp[0] = null;
+                try {
+                    sockp[0] = bindAddress == null ? new Socket(host, port) : new Socket(InetAddress.getByName(host), port, InetAddress.getByName(bindAddress), 0);
```
not sure if null check for the new configuration option would be nicer before creating&setting socket factory. othwerwise new Socket(host,port) will be created which may affect as-is users unless provide unit test case. I guess test case would be nicer to showcase your requirement.

onderson commented on components/camel-ftp/src/main/java/org/apache/camel/component/file/remote/SftpOperations.java:
```diff
@@ -1074,4 +1097,64 @@ public synchronized boolean sendSiteCommand(String command) throws GenericFileOp
+        Socket socket = null;
+        if (timeout == 0) {
+            try {
+                socket = bindAddress == null ? new Socket(host, port) : new Socket(InetAddress.getByName(host), port, InetAddress.getByName(bindAddress), 0);
```
what you do is reasonably good as the class and method has default access modifers.

davsclaus commented:
Thanks for the PR. Is the code where you create a new `Thread` and so some socket validation some code you have seen/copied from JCraft or how do you come up with that?

ffeisst commented:
@davsclaus I have copied the code from jsch.Util.createSocket(..). The comment on the method does also link to the original method. I've done this, to stay as close as possible to the original Socket creation implementation (the original implementation uses Util.createSocket).

@onderson The null check to activate the new behavior only in case of a bind address sounds like a good idea to me. I've changed the commit accordingly. I also like the idea of a test case and already thought about it, but I dont know how I can setup an environment with multiple interfaces to test the bind address behavior in an unit test. Do you know how this can easily be done in an unit test? I've tested the implementation in a real world setup with multiple interfaces and it is working as expected.

onderson commented:
maybe you can write a Unit test against your environment and mark those test case(s) `@Ignored`. (Of course, when you commit, please mask your ip addresses) If it is hard to unit test and we can not use any mocking mechanism, that's what we can usually do. if it is more complicated, you can leave it as is i suppose (in code terms, AFAIU this now looks like a new option without any effect on existing users.)

onderson commented on components/camel-ftp/src/main/java/org/apache/camel/component/file/remote/SftpOperations.java:
```diff
@@ -1074,4 +1099,64 @@ public synchronized boolean sendSiteCommand(String command) throws GenericFileOp
+            tmp.join(timeout);
+            message = "timeout: ";
+        } catch (java.lang.InterruptedException eee) {
+        }
```
Can we at least add here a debug log?

davsclaus commented on components/camel-ftp/src/main/java/org/apache/camel/component/file/remote/SftpOperations.java:
```diff
@@ -1074,4 +1099,64 @@ public synchronized boolean sendSiteCommand(String command) throws GenericFileOp
+            }
+            tmp.interrupt();
+            tmp = null;
+            throw new RuntimeCamelException(message, ee[0]);
```
Should this not be IOException and then let the method have throws IOException

davsclaus commented on components/camel-ftp/src/main/java/org/apache/camel/component/file/remote/SftpConfiguration.java:
```diff
@@ -69,6 +69,8 @@
     private LoggingLevel jschLoggingLevel = LoggingLevel.WARN;
     @UriParam(label = "advanced")
     private Integer bulkRequests;
+    @UriParam(label = "bindAddress")
```
I guess this option is more common, so we can remove the label

dmvolod commented:
@ffeisst in advance to @onderson suggestion, you can try to add bindAddress unit test with camel-testcontainer functiality. If Docker will find on the test env, you can bind to the real address of the chose container.

davsclaus commented:
The unit test would be nice, but as said its a bit hard to do without real containers, so the testcontainer is a possibility. But its not 100% necessary (but would be nice to get more integration tests in camel component with testcontainers), it seems this code is only active if you set the bindAddress option and therefore not in risk of affecting existing behaviour.

davsclaus commented:
I am merging this, then we can do another PR if we manage to build an unit test
